dieharder/dieharder.1
.\" $Id: dieharder.1 318 2007-04-29 18:06:32Z rgb $
.TH DieHarder 1 "Copyright 2004 Robert G. Brown" "DieHarder"
.SH NAME
rand_rate \- A testing and benchmarking tool for GSL random number 
generators

.SH SYNOPSIS
dieharder [-a] [-b bits] [-d diehard test number] [-f filename]
          [-g generator number] [-h] [-i iterations] [-l]
          [-n ntuple] [-p number of p samples] [-q] [-N]
          [-r rgb test number] [-s sts test number]
          [-t number of test samples] [-u user test number]
          [-v verbose flag] [-x xvalue] [-y yvalue] [-z zvalue]

.SH DieHarder OPTIONS

.TP
-a Runs all the tests with standard/default options to create a report
.TP
-b bits - sets the number of bits to be used in tests that act on a bit
string of variable length, e.g. the rgb bitdist test.
.TP
-d test number -  selects specific diehard test.
.TP
-f filename will EVENTUALLY permit random strings to be tested to be
read in from a file, but this is not yet implemented!
.TP
-g generator number - selects a specific generator for testing.  Using
-1 causes all known generators to be printed out to the display.
.TP
-h prints context-sensitive help -- usually Usage (this message) or a
test synopsis if entered as e.g. dieharder -d 3 -h.
.TP
-i iterations - sets iteration count for timing runs (should not be
needed).
.TP
-l list all known tests.
.TP
-n ntuple - set ntuple length for tests on short bit strings that permit
the length to be varied (e.g. rgb bitdist).
.TP
-p count - sets the number of p-value samples per test (default 100).
.TP
-q selects \"quiet\" operation: results only are printed on a single line
(where applicable).  This is a way of getting a very compact report.
.TP
-N selects non-overlapping samples even in tests that \"require\"
overlapping samples, such as operm5.  In some cases this will matter.
In others it will not.
.TP
-r test number - selects specific rgb test.
.TP
-s test number - selects specific sts test.
.TP
-t count - sets the number of random entities used in each test, where
possible.  Be warned -- some tests will take a long time with the
default value of 10000.  Read the test synopses for suggested settings
for -t or use -a first.
.TP
-u test number - selects specific user-developed test, if you've added
one or more to this tool.  dieharder provides this interface to make
it easy to add your own tests.
.TP
-v verbose flag -- controls the verbosity of the output for debugging
only.  Probably of little use to non-developers, and developers can
read the enum(s) in dieharder.h and the test sources to see which
flag values turn on output on which routines.  1 is \"all\" and will
result in a fairly detailed trace of program activity.
.TP
-x,-y,-z number - Some tests have parameters that can safely be varied
from their default value.  For example, in the diehard birthdays test,
one can vary the number of \"dates\" drawn from the \"year\" of some
length, which can also be varied.  -x 2048 -y 30 alters these two
values but should still run fine.

.B NOTE WELL:  
The "bogomegarates" returned by this tool are BOGUS and may not be even 
approximately correct in your context.  Also, the quality assessment(s) 
for the rngs may, in fact, be completely incorrect or misleading.  Use 
them at your Own Risk!  Be Warned!


.SH DESCRIPTION
.B DieHarder
This is the current snapshot of the dieharder random number
tester.  It encapsulates all of the Gnu Scientific Library random number
generators as well as /dev/random and /dev/urandom in a single harness
that can time them and subject them to various tests for randomness.
These tests are variously drawn from George Marsaglia's "Diehard battery of
random number tests", the NIST Statistical Test Suite, and include a
few useful e.g. binning or spectral tests that I've found doing research 
into random number tests or tests that I myself have made up or that are 
improvements on tests from other sources.

The primary point of this tester is to make it easy to time and test
(pseudo)random number generators OR hardware or other generators.  Three
examples are provided of wrapping a random number generator and
inserting it so that it is can be called via the GSL interface.  An
interface is planned that would allow random numbers to be read in from
a file, allowing the tool to be used with any generator (wrappable or
not) that can generate a table of random numbers.

Another important motivation for writing dieharder is to have the
entire test code and documentation be fully Gnu Public Licensed and
hence openly available for adaptation, testing, modification, including
the addition of new tests.  The primary objections I have towards
diehard and sts are not that they are or are not adequate and complete;
it is that the code is obscure and not explicitly published for reuse.
It is my hope that by providing this tool in autodocumenting source, it
makes it easy to add new tests, critique older tests, and improve the
suite in general.

.SH FILE INPUT

To test your own random number generator or to test the randomness of
some file of presumably random numbers, two approaches are possible.
First of all, there are five generators that are wrapped up in GSL
compatible clothes and linked to the GSL so that the standard GSL
interface works for them.  Using these as prototypes and working with
the fully GPL dieharder sources, any software or hardware random number
generator that one wishes to test can be added for testing using these
as prototypes, and can likely be submitted to the GSL for inclusion if
they pass the tests as well or better than the generators that are already
there.  Dieharder is designed to (ultimately) be a
.I very convenient tool
for testing new software RNGs.

However, the last two non-gsl generators are "universal" generators in
the sense that they permit you to input a random number stream from a
.I file 
(but 
.I not 
from /dev/random or /dev/urandom, be warned).  

The file_input generator requires a file of "cooked" (ascii readable)
random numbers, one per line, with a header that describes the format to
dieharder.  This interface is still somewhat experimental -- not all
ascii formats have been tested.  However, it has been tested and should
work for 32 bit unsigned integers represented directly in ascii or as 32
bits of binary.  Example files for a couple of possible input formats
are included in the sources.

Finally, the type file_input_raw accepts a file of raw bits as input,
such as might be generated by:

 dd if=/dev/urandom of=testrands.raw bs=4 count=1000000

(to generate 1,000,000 four-byte ints directly from the
software-augmented kernel entropy generator).  That is, running the
tests from such a file should be 
.I approximately
the same as testing /dev/urandom directly.

The main (important!) difference is that some of the test require a
.I lot
of random numbers -- far more than were needed by diehard.  Indeed,
dieharder typically runs many of the diehard tests 100 
.I independent 
times, generating a p-value for each, and plots a histogram of the
p-values and generates a p-value for the (presumed uniform) distribution
of p-values!  This approach mimics the histogram presented in the STS
suite but augments it with a hard number.

This protects one from the "p happens" problem -- every now and then you
will have a run with a very low or high p from a good generator, but
over time a good generator will generate a uniform distribution of
p-values.  Dieharder lets you visually decide if the distribution is or
isn't credibly uniform, while giving you an index that in most cases is
a fairly clear "good" or "bad" indicator for a given random sequence or
generator.

File input rands are delivered to the tests on demand, but if the test
needs more than are available it simply rewinds the file and cycles
through it again, and again, and again as needed.  Obviously this
significantly reduces the sample space and can lead to completely
incorrect results for the p-value histograms unless there are enough
rands to run EACH test without repetition (it is harmless to reuse the
sequence for different tests).  Let the user beware!

.SH PUBLICATION RULES
.B DieHarder
is entirely original code and can be modified and used at will by any 
user, provided that:

  a) The original copyright notices are maintained and that the source,
including all modifications, is made publically available at the time of
any derived publication.  This is open source software according to the
precepts and spirit of the Gnu Public License.  See the accompanying
file COPYING, which also must accompany any redistribution.

  b) The author of the code (Robert G. Brown) is appropriately
acknowledged and referenced in any derived publication.  It is strongly
suggested that George Marsaglia and the Diehard suite and the various 
authors of the Statistical Test Suite be similarly acknowledged, although 
this suite shares no actual code with these random number test suites.

  c) Full responsibility for the accuracy, suitability, and
effectiveness of the program rests with the users and/or modifiers.  As
is clearly stated in the accompanying copyright.h:

THE COPYRIGHT HOLDERS DISCLAIM ALL WARRANTIES WITH REGARD TO THIS
SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND
FITNESS, IN NO EVENT SHALL THE COPYRIGHT HOLDERS BE LIABLE FOR ANY
SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER
RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

.SH ACKNOWLEDGEMENTS
The author of this suite gratefully acknowledges George Marsaglia (the
author of the diehard test suite) and the various authors of NIST
Special Publication 800-22 (which describes the Statistical Test Suite
for testing pseudorandom number generators for cryptographic
applications), for excellent descriptions of the tests therein.  These
descriptions enabled this suite to be developed with a clean copyright,
licensable under the GPL.

The author also wishes to reiterate that the academic correctness and 
accuracy of the implementation of these tests is his sole responsibility 
and not that of the authors of the Diehard or STS suites.  This is
especially true where he has seen fit to modify those tests from their
strict original descriptions.

.SH COPYRIGHT
GPL 2b; see the file COPYING that accompanies the source of this
program.  This is the "standard Gnu General Public License version 2 or
any later version", with the one minor (humorous) "Beverage"
modification listed below.  Note that this modification is probably not
legally defensible and can be followed really pretty much according to
the honor rule.

As to my personal preferences in beverages, red wine is great, beer is
delightful, and Coca Cola or coffee or tea or even milk acceptable to
those who for religious or personal reasons wish to avoid stressing my
liver.

.B The "Beverage" Modification to the GPL:

Any satisfied user of this software shall, upon meeting the primary
author(s) of this software for the first time under the appropriate
circumstances, offer to buy him or her or them a beverage.  This
beverage may or may not be alcoholic, depending on the personal ethical
and moral views of the offerer.  The beverage cost need not exceed one
U.S. dollar (although it certainly may at the whim of the offerer:-) and
may be accepted or declined with no further obligation on the part of
the offerer.  It is not necessary to repeat the offer after the first
meeting, but it can't hurt...



ChangeLog
2008-08-18 19:50  rgb

	* .: This might be ready for a production snapshot, or a testing
	  snapshot,
	  or something like that. At any rate, I have BOTH fixed the
	  generator
	  problem (where rngs have numbers that can change as new
	  generators
	  are added. I fixed this by significantly increasing the size of
	  the
	  types[] vector and working moderately hard to pack it with
	  generators
	  sparsely and sanely -- sparse so there is plenty of room, sane
	  because
	  the space is now blocked out by general category with room to
	  grown.
	  
	  This will NOT GUARANTEE that the GSL will not change its internal
	  order,
	  and dieharder will continue to follow the GSL order where they
	  overlap
	  as nothing else makes sense.
	  
	  I >>also<< think I've fixed -r 5, rgb_permutations. This is a
	  non-overlapping permutations test and runs to more or less
	  arbitrary
	  order (and seems to work quite nicely where it runs, although it
	  gets
	  quite slow when we get to very large permutations as one might
	  expect).
	  
	  I think it is time to bump the minor-minor number to 11 and post
	  it, for
	  the time being with -r 5 still marked out as experimental in the
	  help listing.

2008-08-18 18:33  rgb

	* .: This fixes the only real bug, I think -- dieharder now dies
	  with a
	  complaint if it is invoked with an invalid -g entry, sort of.
	  Hmmm,
	  actually I might need to tweak this fix just a bit more to check
	  on the complete range.

2008-08-18 18:29  rgb

	* .: This actually works and has the number spaces previously
	  discussed
	  all implemented and everything.

2008-08-18 00:15  rgb

	* .: This is a step in the process of figuring out how to "fix" the
	  numbers
	  for rngs to test. I think I'm going to have to move my own setup
	  routine to libdieharder. But I'm still trying to figure out
	  whether or
	  not I can or should replace the gsl code itself, or run something
	  in
	  parallel.

2008-08-17 17:30  rgb

	* .: This rearranges the Bugs stuff. I need to create a "cleared"
	  directory
	  in Bugs so I can see what is done, as well.

2008-08-17 16:31  rgb

	* .: This updates the abstract, which is also important
	  documentation and has
	  been added to the %doc lines in the specfile. I should really
	  create a
	  make installdoc target in Makefile.am that can install a bunch of
	  documents, but that is for another day.

2008-08-15 23:03  rgb

	* .: This is 2.27.10-1, and adds stdin support and fixes a few bugs
	  in the
	  specfile. I'm posting it for distribution, although it is still
	  not
	  "finished" -- this is very much development snapshot time.

2008-08-15 18:14  rgb

	* .: This adds the capability to input a binary stream into stdin.
	  This
	  patch is due to Paul Serice. He also included a (standalone?) rng
	  to
	  test with, which I'll mess with in a moment.

2008-08-03 12:40  rgb

	* .: Added a version string to help. Probably should add one
	  elsewhere, e.g.
	  at the top of rng listing and test listing.

2008-08-01 13:11  rgb

	* .: Had to back off the svn2cl in the rpm build -- it now has to
	  be run BY
	  HAND before just MY rpmbuild (dummy me) so that it won't fail for
	  end
	  users without svn2cl and an svn repo to match. It still builds
	  perfectly, and I'll do a full rpm build cycle including this step
	  to get
	  this last comment into the changelog.

2008-07-29 16:01  rgb

	* .: ...And it is. So henceforth ALL of my svn checkins will be
	  duly noted
	  in the ChangeLog, which will therefore get to be very long
	  indeed, as I
	  check in a lot.
	  
	  Damn! I guess this means no more swearing, and I sure hope I
	  haven't
	  impugned anyone's character in the first 400 plus checkins. At
	  least
	  not too much. Hmmm, I wonder just what I HAVE said down there. I
	  tend
	  to treat SVN logs as sort of a free-association commentary about
	  the
	  project in addition to being a place to document changes. Some
	  changes
	  have "real" documentation in the form of the svn diffs, as well
	  -- I
	  haven't been too specific about just EXACTLY what I've changed
	  every
	  time.
	  
	  Well, better a ChangeLog up to date and verbose than one months
	  to years
	  old, and I'll be hogtied and hung by the nose before I write a
	  precisely
	  formatted description of changes day to day as I work on this --
	  twice.
	  So SVN it is (dating back to when it was really CVS, BTW,
	  imported to
	  SVN).
	  
	  Next, I suppose we should provide at least >>a<< fix for the
	  problem of
	  shifting generator numbers, and on my own I want to clean up some
	  issues with file read -- arrange it so we can read in very large
	  files
	  and eliminate/document file-based usage redundancies such as
	  needing -g
	  65 AND -f filename. This will be a bit tricky, as dieharder can
	  read in
	  raw (binary) as well as ascii in many formats, and using the -g
	  flag to
	  differentiate at least raw vs cooked has been an easy solution.
	  
	  Regarding numbers, it has been suggested that we do two things:
	  
	  a) Take action to ensure that the numbers don't change any more,
	  even if
	  the gsl or I or a user add new generators. This is actually a bit
	  of a
	  PITA (much as I see the motivation for doing so, as moving
	  numbers break
	  user scripts) because the way generators are added is to run
	  through all
	  the supported ones of each type and just add them in gsl or call
	  order.
	  However, there are several ways I can do it (some requiring more
	  programming than others) and I'll see which one looks to be
	  easiest and
	  most robust.
	  
	  b) Arrange it so we can call generators by name, not just by
	  number.
	  Names would presumably not change even if numbers were inserted,
	  and
	  many users will be interested in testing a single generator they
	  are
	  thinking of using by name.
	  
	  rgb

2008-07-29 15:40  rgb

	* .: OK, this seems to work, autobuilding ChangeLog from the svn
	  log of my

RDieHarder/R/dieharder.R

.dieharder.generators <- c("borosh13",
                           "cmrg",
                           "coveyou",
                           "fishman18",
                           "fishman20",
                           "fishman2x",
                           "gfsr4",
                           "knuthran",
                           "knuthran2",
                           "knuthran2002",
                           "lecuyer21",
                           "minstd",
                           "mrg",
                           "mt19937",
                           "mt19937_1999",
                           "mt19937_1998",
                           "r250",
                           "ran0",
                           "ran1",
                           "ran2",
                           "ran3",
                           "rand",
                           "rand48",
                           "random128-bsd",
                           "random128-glibc2",
                           "random128-libc5",
                           "random256-bsd",
                           "random256-glibc2",
                           "random256-libc5",
                           "random32-bsd",
                           "random32-glibc2",
                           "random32-libc5",
                           "random64-bsd",
                           "random64-glibc2",
                           "random64-libc5",
                           "random8-bsd",
                           "random8-glibc2",
                           "random8-libc5",
                           "random-bsd",
                           "random-glibc2",
                           "random-libc5",
                           "randu",
                           "ranf",
                           "ranlux",
                           "ranlux389",
                           "ranlxd1",
                           "ranlxd2",
                           "ranlxs0",
                           "ranlxs1",
                           "ranlxs2",
                           "ranmar",
                           "slatec",
                           "taus",
                           "taus2",
                           "taus113",
                           "transputer",
                           "tt800",
                           "uni",
                           "uni32",
                           "vax",
                           "waterman14",
                           "zuf",
                           "dev/random",
                           "dev/urandom",
                           "empty",
                           "file_input",
                           "file_input_raw",
                           "ca",
                           "R_wichmann_hill",
                           "R_marsaglia_multic.",
                           "R_super_duper",
                           "R_mersenne_twister",
                           "R_knuth_taocp",
                           "R_knuth_taocp2"
                           )

dieharderGenerators <- function() {
  val <- .Call("dieharderGenerators", PACKAGE="RDieHarder")
  return(val[[1]])
}

.dieharder.tests <- c("birthdays",
                      "operm5",
                      "rank_32x32",
                      "rank_6x8",
                      "bitstream",
                      "opso",
                      "oqso",
                      "dna",
                      "count_1s_stream",
                      "count_1s_byte",
                      "parking_lot",
                      "2dsphere",
                      "3dsphere",
                      "squeeze",
                      "sums",
                      "runs",
                      "craps",
                      "marsaglia_tsang_gcd",
                      "marsaglia_tsang_gorilla",
                      "rgb_timing",
                      "rgb_persist",
                      "rgb_bitdist",
                      "rgb_lmn",
                      "sts_monobit",
                      "sts_runs",
                      "sts_block")
 
dieharder <- function(rng="mt19937",
                      test=1,
                      psamples=100,
                      seed=0,
                      verbose=FALSE) {
  UseMethod("dieharder")
}

dieharder.default <- function(rng="mt19937",
                              test="runs",
                              psamples=100,
                              seed=0,
                              verbose=FALSE) {

  if (length(rng) > 1) {
    warning("Only one rng argument supported in dieharder")
  }
  if (is.character(rng)) {
    ## dieharder's C code indexes these from 0 to 73, not 1 to 74
    gen <- charmatch(rng, .dieharder.generators) - 1
  } else {
    gen <- rng - 1
  }
  if (is.na(gen)) {
    warning("rng argumement ", rng, " unknown")
    return(NULL)
  }
  if (gen == 0 && rng != 1) {
    warning("rng argumment ", rng, " ambiguous")
    return(NULL)
  }

  if (length(test) > 1) {
    warning("Only one test argument supported in dieharder")
  }
  if (is.character(test)) {
    ## note that dieharder uses a 0-based C array, so subtract one
    runtest <- charmatch(test, .dieharder.tests) 
  } else {
    runtest <- test 
  }
  if (is.na(runtest)) {
    warning("test argumement ", test, " unknown")
    return(NULL)
  }
  if (runtest == 0) {
    warning("test argumment ", test, " ambiguous")
    return(NULL)
  }

  
  val <- .Call("dieharder",
               as.integer(gen),
               as.integer(runtest),
               as.integer(seed),
               as.integer(psamples),
               as.integer(verbose),
               PACKAGE="RDieHarder")
  ##class(val) <- c("dieharder", "test")
  obj <- list(p.value=val[[1]],
              data=val[[2]],      ## not used bu htest methods
              method=val[[3]],
              data.name=paste("Created by RNG `",
                .dieharder.generators[gen+1], "' with seed=",
                as.integer(seed), ", sample of size ",
                as.integer(psamples), sep=""),
              generator=.dieharder.generators[gen+1]
              )
  class(obj) <- c("dieharder", "htest")
  return(obj)
}

plot.dieharder <- function(x, ...) {

  local.par <- par(mfrow=c(2,1), mar=c(2,4,3,1), oma=c(0,0,2.5,0))

  vec <- x$data

  hist(vec, probability=TRUE, main="Histogram and Density estimate",
       xlab="", ylab="density")
  lines(density(x$data, from=0, to=1), col='darkgray')

  #qqplot(vec, seq(0, 1, length.out=length(vec)),
  #       main="QQ-Plot", ylab="Uniform sequence", xlab="")

  plot(ecdf(x$data), main="ECDF", ylab="", xlab="",
       verticals= TRUE, do.p = FALSE)
  segments(0,0,1,1, col='darkgray', lty="dotted")

  mtext(text = paste(x$method, "applied to data generated by", 
          x$generator), outer = TRUE, cex = 1.3, font = 2, line = 1)
  pksk <- round(x$p.value, 4)
  pks <- round(ks.test(x$data, "punif", 0, 1, exact=TRUE)$p.value, 4)
  pw <- round(wilcox.test(x$data, mu=0.5)$p.value, 4)
  mtext(text = paste("Test p-values: ",
          pksk, " (Kuiper-K-S),  ",
          pks, " (K-S),  ",
          pw, "  (Wilcoxon)"),
        outer = TRUE, cex = 1.0, font = 1, line = 0)
  
  par(local.par)
  invisible(x)
}

print.dieharder <- function(x, ...) {
  stats:::print.htest(x, ...)     # why do I need to explicitly dispatch?
  invisible(x)
}

summary.dieharder <- function(object, ...) {
  print(object, ...)
  cat("\nSummary for test data\n")
  print(summary(object$data))
  cat("\n\nStem and leaf plot for test data\n")
  print(stem(object$data))
  print(ks.test(object$data, "punif", 0, 1, exact=TRUE))
  print(wilcox.test(object$data, mu=0.5))
  invisible(object)
}

##

#dieharderVec <- function(gen=c(22,69), test=1, verbose=FALSE) {
#
#  val <- .Call("dieharderVectorised",
#                as.list(as.integer(gen)),
#                as.integer(test),
#                as.integer(verbose),
#                PACKAGE="RDieHarder")
#   return(val)
#}


